### PR TITLE
Hardhat allows external artifacts

### PR DESCRIPTION
**Pull request by ducquangkstn**
Resolve https://github.com/ethereum-ts/TypeChain/issues/367

### CODE REVIEW

File: packages/hardhat/src/index.ts [hunk 1]
```diff
@@ -43,15 +44,19 @@ subtask(TASK_COMPILE_SOLIDITY_COMPILE_JOBS, 'Compiles the entire project, buildi
...
-    const needsFullRebuild = taskArgsStore.fullRebuild || typechainCfg.target !== 'ethers-v5'
+    const needsFullRebuild =
+      taskArgsStore.fullRebuild || typechainCfg.target !== 'ethers-v5' || typechainCfg.externalArtifacts
     console.log(
       `Generating typings for: ${artifactPaths.length} artifacts in dir: ${typechainCfg.outDir} for target: ${typechainCfg.target}`,
     )
     const cwd = process.cwd()
-    const { runTypeChain, glob } = await import('typechain')
-    const allFiles = glob(cwd, [`${config.paths.artifacts}/!(build-info)/**/+([a-zA-Z0-9_]).json`])
+    const artifactsPaths = [`${config.paths.artifacts}/!(build-info)/**/+([a-zA-Z0-9_]).json`]
```
Comment (krzkaczor) [inline packages/hardhat/src/index.ts:55]: I think the better implementation would be to first run glob ignoring node_modules and later again to get external artifacts. Finally, merge both results. This way we don't modify old behaviour at all.

File: packages/hardhat/src/index.ts [hunk 2]
```diff
@@ -43,15 +44,19 @@ subtask(TASK_COMPILE_SOLIDITY_COMPILE_JOBS, 'Compiles the entire project, buildi
 
     // RUN TYPECHAIN TASK
     const typechainCfg = config.typechain
-    // incremental generation is only supported in 'ethers-v5'
+    // incremental generation is only supported in 'ethers-v5' and external artifacts
     // @todo: probably targets should specify somehow if then support incremental generation this won't work with custom targets
-    const needsFullRebuild = taskArgsStore.fullRebuild || typechainCfg.target !== 'ethers-v5'
+    const needsFullRebuild =
```
Comment (krzkaczor) [inline packages/hardhat/src/index.ts:49]: Can we somehow support incremental generation with `externalArtifacts`? If I get it correctly we have no way of checking if we already generated types for externalArtifacts and this is the reason for this change. But maybe we can always generate typings only for `externalArtifacts` and avoid regenerating all types for internal files? Incremental generation is often a lifesaver.

**Review by krzkaczor — CHANGES_REQUESTED** [review]
Thanks for working on it! Some comments.

**Comment by ducquangkstn** [comment]
@krzkaczor can you review this PR again?

**Review by krzkaczor — APPROVED** [review]

**Comment by krzkaczor** [comment]
Thanks! I will add changesets and try to release it soon.

**Comment by krzkaczor** [comment]
Released as `@typechain/hardhat@2.1.0`

**Pull request merged**